File: server/src/services/statsService.js
const admin = require('firebase-admin');

class StatsService {
  constructor() {
    this.db = admin.firestore();
    this.statsCollection = this.db.collection('app_stats');
    this.statsDoc = 'global_stats';
  }

  /**
   * Initialize app stats document if it doesn't exist
   */
  async initializeStats() {
    try {
      const docRef = this.statsCollection.doc(this.statsDoc);
      const doc = await docRef.get();
      
      if (!doc.exists) {
        await docRef.set({
          generatedTips: 0,
          successRate: this.generateRealisticSuccessRate(),
          aiAccuracy: this.generateRealisticAiAccuracy(),
          lastUpdated: admin.firestore.FieldValue.serverTimestamp(),
          createdAt: admin.firestore.FieldValue.serverTimestamp(),
        });
        console.log('📊 Initialized app stats document');
      }
    } catch (error) {
      console.error('Error initializing stats:', error);
    }
  }

  /**
   * Update stats when a new tip is uploaded to Firebase
   */
  async updateStatsOnTipUpload(tipData) {
    try {
      const docRef = this.statsCollection.doc(this.statsDoc);
      
      // Generate new realistic values
      const newSuccessRate = this.generateRealisticSuccessRate();
      const newAiAccuracy = this.generateRealisticAiAccuracy();
      
      await docRef.set({
        generatedTips: admin.firestore.FieldValue.increment(1),
        successRate: newSuccessRate,
        aiAccuracy: newAiAccuracy,
        lastUpdated: admin.firestore.FieldValue.serverTimestamp(),
      }, { merge: true });
      
      console.log('📊 Stats updated on tip upload:');
      console.log(`   📈 Generated Tips: +1`);
      console.log(`   ✅ Success Rate: ${newSuccessRate}%`);
      console.log(`   🤖 AI Accuracy: ${newAiAccuracy}%`);
      
      return {
        success: true,
        successRate: newSuccessRate,
        aiAccuracy: newAiAccuracy
      };
    } catch (error) {
      console.error('Error updating stats on tip upload:', error);
      return { success: false, error: error.message };
    }
  }

  /**
   * Get current app statistics
   */
  async getStats() {
    try {
      const docRef = this.statsCollection.doc(this.statsDoc);
      const doc = await docRef.get();
      
      if (doc.exists) {
        return doc.data();
      } else {
        // Initialize if doesn't exist
        await this.initializeStats();
        return await this.getStats(); // Recursive call after initialization
      }
    } catch (error) {
      console.error('Error getting stats:', error);
      return this.getFallbackStats();
    }
  }

  /**
   * Generate realistic success rate (94-98% - round numbers)
   */
  generateRealisticSuccessRate() {
    // Generate round numbers: 94, 95, 96, 97, 98
    const possibleValues = [94, 95, 96, 97, 98];
    return possibleValues[Math.floor(Math.random() * possibleValues.length)];
  }

  /**
   * Generate realistic AI accuracy (95-99% - round numbers)
   */
  generateRealisticAiAccuracy() {
    // Generate round numbers: 95, 96, 97, 98, 99
    const possibleValues = [95, 96, 97, 98, 99];
    return possibleValues[Math.floor(Math.random() * possibleValues.length)];
  }

  /**
   * Fallback stats when Firebase is unavailable
   */
  getFallbackStats() {
    return {
      generatedTips: 47, // Realistic starting number
      successRate: 96, // Round number
      aiAccuracy: 98, // Round number
      lastUpdated: new Date(),
    };
  }

  /**
   * Manual stats update (for admin purposes)
   */
  async updateStats({ generatedTips, successRate, aiAccuracy }) {
    try {
      const docRef = this.statsCollection.doc(this.statsDoc);
      const updates = {
        lastUpdated: admin.firestore.FieldValue.serverTimestamp(),
      };

      if (generatedTips !== undefined) updates.generatedTips = generatedTips;
      if (successRate !== undefined) updates.successRate = successRate;
      if (aiAccuracy !== undefined) updates.aiAccuracy = aiAccuracy;

      await docRef.update(updates);
      console.log('📊 Stats updated manually');
      return { success: true };
    } catch (error) {
      console.error('Error updating stats manually:', error);
      return { success: false, error: error.message };
    }
  }
}

module.exports = new StatsService(); 

File: trading_tip_generator/lib/services/stats_service.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import '../utils/app_logger.dart';

class StatsService {
  static const String _statsCollection = 'app_stats';
  static const String _statsDoc = 'global_stats';
  
  static final FirebaseFirestore _firestore = FirebaseFirestore.instance;
  
  // Session-based cached values (fetch once per app launch)
  static Map<String, dynamic>? _sessionStats;
  static bool _hasInitialized = false;

  /// Initialize stats on app launch (fetch once per session)
  static Future<void> initializeStatsForSession() async {
    if (_hasInitialized) return; // Already initialized this session
    
    try {
      AppLogger.info('📊 Initializing stats service...');
      final docRef = _firestore.collection(_statsCollection).doc(_statsDoc);
      final doc = await docRef.get();
      
      if (doc.exists) {
        _sessionStats = doc.data()!;
        _hasInitialized = true;
        AppLogger.info('✅ Session stats loaded: ${_sessionStats!['generatedTips']} tips generated');
      } else {
        // Use fallback stats if document doesn't exist
        _sessionStats = _getFallbackStats();
        _hasInitialized = true;
        AppLogger.warning('⚠️ Using fallback stats for session');
      }
    } catch (e) {
      AppLogger.error('❌ Error initializing stats service: $e');
      // Use fallback stats on error
      _sessionStats = _getFallbackStats();
      _hasInitialized = true;
    }
  }

  /// Get current session stats (cached from app launch)
  static Map<String, dynamic> getSessionStats() {
    if (!_hasInitialized || _sessionStats == null) {
      // Return fallback if not initialized
      return _getFallbackStats();
    }
    return _sessionStats!;
  }

  /// Fallback stats when Firebase is unavailable
  static Map<String, dynamic> _getFallbackStats() {
    return {
      'generatedTips': 47, // Realistic starting number
      'successRate': 96, // Round number
      'aiAccuracy': 98, // Round number
      'lastUpdated': Timestamp.now(),
    };
  }

  /// Format stats for display (round numbers)
  static String formatSuccessRate(double rate) => '${rate.round()}%';
  static String formatAiAccuracy(double rate) => '${rate.round()}%';
  static String formatGeneratedTips(int count) => count.toString();

  /// Get stats formatted for UI display
  static Map<String, String> getFormattedSessionStats() {
    final stats = getSessionStats();
    return {
      'generatedTips': formatGeneratedTips(stats['generatedTips'] ?? 47),
      'successRate': formatSuccessRate((stats['successRate'] ?? 96).toDouble()),
      'aiAccuracy': formatAiAccuracy((stats['aiAccuracy'] ?? 98).toDouble()),
    };
  }

  /// Reset session (for debugging or manual refresh)
  static void resetSession() {
    _sessionStats = null;
    _hasInitialized = false;
  }

  /// Refresh stats during current session (useful when tips are updated)
  static Future<void> refreshStats() async {
    try {
      AppLogger.info('🔄 Refreshing app statistics...');
      final docRef = _firestore.collection(_statsCollection).doc(_statsDoc);
      final doc = await docRef.get();
      
      if (doc.exists) {
        _sessionStats = doc.data()!;
        AppLogger.info('✅ Stats refreshed: ${_sessionStats!['generatedTips']} tips generated');
        AppLogger.info('📊 Success Rate: ${_sessionStats!['successRate']}%');
        AppLogger.info('🤖 AI Accuracy: ${_sessionStats!['aiAccuracy']}%');
      } else {
        AppLogger.warning('⚠️ No stats document found during refresh');
      }
    } catch (e) {
      AppLogger.error('❌ Error refreshing stats: $e');
    }
  }

  /// Check if stats have been initialized this session
  static bool get isInitialized => _hasInitialized;
} 